### PR TITLE
Adding the Profiling capability for grpc-backend and http-gateway. 

### PR DESCRIPTION
**Pull request by anichawla**
Adding the Profiling capability for grpc-backend and http-gateway and
updating README for instructions to start the profiling services

### CODE REVIEW

File: cmds/http-gateway/main.go [hunk 1]
```diff
@@ -224,8 +226,17 @@ func main() {
 		logger = logging.WithValuesFromContext(ctx, logging.Logger)
 	)
 
+	if *profServiceName != "" {
+		if err := profiler.Start(profiler.Config{
+			Service:        *profServiceName,
+			ServiceVersion: "v1"}); err != nil {
```
Comment (steeling) [inline cmds/http-gateway/main.go]: let's put this on 2 lines for readability

File: cmds/grpc-backend/main.go [hunk 2]
```diff
@@ -173,6 +175,13 @@ func main() {
 		ctx    = context.Background()
 		logger = logging.WithValuesFromContext(ctx, logging.Logger)
 	)
+	if *profServiceName != "" {
+		if err := profiler.Start(profiler.Config{
+			Service:        *profServiceName,
+			ServiceVersion: "v1"}); err != nil {
```
Comment (steeling) [inline cmds/grpc-backend/main.go]: same comment about 2 lines as below

File: cmds/grpc-backend/main.go [hunk 3]
```diff
@@ -34,6 +35,7 @@ var (
 	logFormat         = flag.String("log_format", logging.DefaultFormat, "The log format in {json, console}")
 	logLevel          = flag.String("log_level", logging.DefaultLevel.String(), "The log level")
 	dumpRequests      = flag.Bool("dump_requests", false, "Log request and response protos")
+	profServiceName   = flag.String("prof_service_name", "", "Service name for the Go profiler")
```
Comment (steeling) [inline cmds/grpc-backend/main.go]: lets prefix this flag with `gcp_`

File: cmds/http-gateway/main.go [hunk 4]
```diff
@@ -24,9 +25,10 @@ import (
 )
 
 var (
-	address       = flag.String("addr", ":8080", "Local address that the gateway binds to and listens on for incoming connections")
-	traceRequests = flag.Bool("trace-requests", false, "Logs HTTP request/response pairs to stderr if true")
-	grpcBackend   = flag.String("grpc-backend", "", "Endpoint for grpc backend. Only to be set if run in proxy mode")
+	address         = flag.String("addr", ":8080", "Local address that the gateway binds to and listens on for incoming connections")
+	traceRequests   = flag.Bool("trace-requests", false, "Logs HTTP request/response pairs to stderr if true")
+	grpcBackend     = flag.String("grpc-backend", "", "Endpoint for grpc backend. Only to be set if run in proxy mode")
+	profServiceName = flag.String("prof_service_name", "", "Service name for the Go profiler")
```
Comment (steeling) [inline cmds/http-gateway/main.go]: also prefix with gcp_

**Comment by steeling** [comment]
@vosst thoughts on these changes?

File: cmds/http-gateway/main.go (hunk 1 above)
Comment (anichawla) [reply cmds/http-gateway/main.go]: Done

File: cmds/grpc-backend/main.go (hunk 2 above)
Comment (anichawla) [reply cmds/grpc-backend/main.go]: Done

File: cmds/grpc-backend/main.go (hunk 3 above)
Comment (anichawla) [reply cmds/grpc-backend/main.go]: Done

File: cmds/http-gateway/main.go (hunk 4 above)
Comment (anichawla) [reply cmds/http-gateway/main.go]: Done

File: cmds/grpc-backend/main.go [hunk 5]
```diff
@@ -173,6 +175,14 @@ func main() {
 		ctx    = context.Background()
 		logger = logging.WithValuesFromContext(ctx, logging.Logger)
 	)
+	if *profServiceName != "" {
+		if err := profiler.Start(
+			profiler.Config{
+				Service:        *profServiceName,
+				ServiceVersion: "v1"}); err != nil {
+			logger.Error("Failed to start the profiler ", zap.Error(err))
```
Comment (vosst) [inline cmds/grpc-backend/main.go]: Seems like we should rather be strict here and use `logger.Panic`, instead. With that, any issues with setting up the profiling become very visible.

File: cmds/grpc-backend/main.go [hunk 6]
```diff
@@ -173,6 +175,14 @@ func main() {
 		ctx    = context.Background()
 		logger = logging.WithValuesFromContext(ctx, logging.Logger)
 	)
+	if *profServiceName != "" {
+		if err := profiler.Start(
+			profiler.Config{
+				Service:        *profServiceName,
+				ServiceVersion: "v1"}); err != nil {
```
Comment (vosst) [inline cmds/grpc-backend/main.go]: Seems like we should factor out the service version into a named constant.

File: cmds/http-gateway/main.go (hunk 1 above)
Comment (steeling) [reply cmds/http-gateway/main.go]: I mean doing this:

err := profiler.Start(...)

if err !=nil {

}

A multiline if err := ... ; err != nil is a little cumbersome

Comment (anichawla) [reply cmds/http-gateway/main.go]: Done

File: cmds/grpc-backend/main.go (hunk 6 above)
Comment (anichawla) [reply cmds/grpc-backend/main.go]: Done.

File: cmds/grpc-backend/main.go (hunk 5 above)
Comment (anichawla) [reply cmds/grpc-backend/main.go]: Done..

**Comment by steeling** [comment]
azure/pipelines

**Pull request merged**